### PR TITLE
12 create assignment

### PR DESCRIPTION
**Pull request by kongebra**
### Added
* Create Assignment
* Create Submission Form
* Form.js -> a small framework for creating small custom forms
* Created relationized database tables for assignments, submissions, reviews, and the custom forms

### CODE REVIEW

File: controller/admin.go [hunk 1]
```diff
@@ -188,6 +199,187 @@ func AdminAssignmentGET(w http.ResponseWriter, r *http.Request) {
...
+	success, err := assignmentRepository.Insert(assignment)
+	if err != nil {
+		log.Println(err)
+		return
+	}
+
+	if success {
+		// TODO (Svein): Celebrate
+	}
+
```
Comment (JohanAanesen) [inline controller/admin.go]: mangler redirection
http.Redirect(w, r, "/", http.StatusFound)

File: controller/admin.go [hunk 2]
```diff
@@ -188,6 +199,187 @@ func AdminAssignmentGET(w http.ResponseWriter, r *http.Request) {
...
+	var repo = model.SubmissionRepository{}
+
+	err = repo.Insert(form)
+	if err != nil {
+		log.Println(err)
+		return
+	}
+
+	// TODO (Svein): Redirect or something!?
```
Comment (JohanAanesen) [inline controller/admin.go]: redirect
http.Redirect(w, r, "/", http.StatusFound)

File: config/database.sql [hunk 3]
```diff
@@ -16,225 +16,249 @@ USE cs53;
...
-       (3, 'Ola Nordmann', 'olanor@stud.ntnu.no', 1, 'swag-meister69@ggmail.com',
-        '$2a$14$vH/ibjwwXqBmOgJt8JCiK.S7D2r0VrBu46pYdCLs/dJMMk1aBV8RC'), -- Password is 123abc --
-       (4, 'Johan Klausen', 'johkl@stu.ntnu.no', 0, NULL,
-        '$2a$14$vH/ibjwwXqBmOgJt8JCiK.S7D2r0VrBu46pYdCLs/dJMMk1aBV8RC'); -- Password is 123abc --
+--
+-- Dumping data for table `users`
+--
 
+INSERT INTO `users` (`id`, `name`, `email_student`, `teacher`, `email_private`, `password`) VALUES
+(2, 'Frode Haug', 'frodehg@teach.ntnu.no', 1, NULL, '$2a$14$vH/ibjwwXqBmOgJt8JCiK.S7D2r0VrBu46pYdCLs/dJMMk1aBV8RC'),
```
Comment (BredeFK) [inline config/database.sql]:
```suggestion
(1, 'Test User', 'hei@gmail.com', 1, 'test@yahoo.com',
        '$2a$14$MZj24p41j2NNGn6JDsQi0OsDb56.0LcfrIdgjE6WmZzp58O6V/VhK'),
       (2, 'Frode Haug', 'frodehg@teach.ntnu.no', 1, NULL,
        '$2a$14$vH/ibjwwXqBmOgJt8JCiK.S7D2r0VrBu46pYdCLs/dJMMk1aBV8RC'), -- Password is 123abc --
```

File: controller/admin.go [hunk 4]
```diff
@@ -91,6 +101,7 @@ func AdminCreateCoursePOST(w http.ResponseWriter, r *http.Request) {
 		Semester:    r.FormValue("semester"),
 	}
 
+	// TODO (Svein): Move this to model, in a function
 	//insert into database
 	result, err := db.GetDB().Exec("INSERT INTO course(hash, coursecode, coursename, year, semester, description, teacher) VALUES(?, ?, ?, ?, ?, ?, ?)",
```
Comment (JohanAanesen) [inline controller/admin.go]: 'hash' unrecognized, also move this into model

File: controller/admin.go [hunk 5]
```diff
@@ -188,6 +199,187 @@ func AdminAssignmentGET(w http.ResponseWriter, r *http.Request) {
...
+		log.Println(err)
+		return
+	}
+
+	deadline, err := DatetimeLocalToRFC3339(r.FormValue("deadline"))
+	if err != nil {
+		log.Println(err)
+		return
+	}
+
```
Comment (BredeFK) [inline controller/admin.go]: Something like: 
```suggestion
if deadline > publish{
    log.println("error")
    return
}
```
Don't let deadline be before publish date

Comment (BredeFK) [reply controller/admin.go]: also use ErrorHandler() more so end user can see somethings wrong

File: controller/admin.go (hunk 4 above)
Comment (JohanAanesen) [reply controller/admin.go]: GetCoursesToUser() fails because hash is nullptr

File: controller/admin.go [hunk 6]
```diff
@@ -188,6 +199,187 @@ func AdminAssignmentGET(w http.ResponseWriter, r *http.Request) {
...
+	w.Header().Set("Content-Type", "text/html; charset=utf-8")
+	w.WriteHeader(http.StatusOK)
+
+	v := view.New(r)
+	v.Name = "admin/assignment/create"
+
+	v.Vars["Courses"] = []struct { // TODO (Svein): use database
```
Comment (BredeFK) [inline controller/admin.go]: Use finished function to get courses
```suggestion

	v.Vars["Courses"] = model.GetCoursesToUser(session.GetUserFromSession(r).ID)
```

File: config/database.sql [hunk 7]
```diff
@@ -16,225 +16,249 @@ USE cs53;
...
+  `year` int(11) NOT NULL,
+  `semester` enum('fall','spring') COLLATE utf8_danish_ci NOT NULL
+) ENGINE=InnoDB DEFAULT CHARSET=utf8 COLLATE=utf8_danish_ci;
+
+--
+-- Dumping data for table `course`
+--
+
+INSERT INTO `course` (`id`, `coursecode`, `coursename`, `teacher`, `description`, `year`, `semester`) VALUES
+(1, 'IMT1031', 'Grunnleggende Programmering', 2, 'Write hello, world in C++', 2019, 'fall'),
```
[the remark this answers is not in the document]
Comment (kongebra) [reply config/database.sql]: ae6faa3 fixed this

File: config/database.sql [hunk 8]
```diff
@@ -16,225 +16,249 @@ USE cs53;
...
 
-CREATE TABLE `course`
-(
-  `id`          int(11)                                        NOT NULL,
-  `hash`        varchar(64)                                    NOT NULL,
-  `coursecode`  varchar(10) COLLATE utf8_danish_ci             NOT NULL,
-  `coursename`  varchar(64) COLLATE utf8_danish_ci             NOT NULL,
-  `teacher`     int(11)                                        NOT NULL,
+CREATE TABLE `course` (
+  `id` int(11) NOT NULL,
```
[the remark this answers is not in the document]
Comment (kongebra) [reply config/database.sql]: ae6faa3

File: config/database.sql [hunk 9]
```diff
@@ -16,225 +16,249 @@ USE cs53;
 SET SQL_MODE = "NO_AUTO_VALUE_ON_ZERO";
 SET AUTOCOMMIT = 0;
 START TRANSACTION;
-SET time_zone = "+01:00"; -- Norwegian time zone! --
+SET time_zone = "+00:00";
```
[the remark this answers is not in the document]
Comment (kongebra) [reply config/database.sql]: ae6faa3

**Comment by kongebra** [comment]
Refactoring request changes done

@JohanAanesen @BredeFK

**Comment by JohanAanesen** [comment]
Some of the visual stuff in /admin is still very off, but i reccon those are quick fixes and can be implemented 'fortløpende' in any current branch

Good work 👍 💦 🇳🇴 🥇

**Pull request merged**